### PR TITLE
Add the Vocab and VocabBuilder data types.

### PR DESCRIPTION
**Pull request by danieldk**
Vocab stores the word embedding vocabulary, including the indexes into
embedding matrices.

VocabBuilder can be used to construct a Vocab, it counts the tokens and
applies the minimum count.

### CODE REVIEW

File: finalfrontier/src/vocab.rs [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                        config.max_n as usize,
+                        config.buckets_exp as usize,
+                    )
+                    .into_iter()
+                    .map(|idx| idx + tokens.len() as u64)
+                    .collect(),
+            );
+        }
+
+        // Invariant: we should have the same number of subwords as token.
```
Comment (DiveFish) [inline finalfrontier/src/vocab.rs]: Typo: [...] tokens.
The comment by itself I find a bit misleading. First, I thought this meant "the number of subwords and the number of tokens should be the same" - though tokens of different length will have a different number of subwords. Yet, it becomes clear with the code it refers to.

File: finalfrontier/src/vocab.rs [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        }
+
+        // Invariant: The index size should be the same as the number of
+        // tokens.
+        assert_eq!(tokens.len(), token_indices.len());
+
+        token_indices
+    }
+
+    /// Get the vocabularly size.
```
Comment (DiveFish) [inline finalfrontier/src/vocab.rs]: vocabulary

File: finalfrontier/src/vocab.rs [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+    /// Get all tokens in the vocabulary.
+    pub fn tokens(&self) -> &[Token] {
+        &self.tokens
+    }
+}
+
+/// This builder is used to construct a vocabulary.
+///
+/// Tokens are added to the vocabulary and counted using the `count` method.
+/// The final vocabulary is contructed using `build`.
```
Comment (DiveFish) [inline finalfrontier/src/vocab.rs]: constructed

File: finalfrontier/src/vocab.rs (hunk 1 above)
Comment (danieldk) [reply finalfrontier/src/vocab.rs]: I think it may be better to remove the comment completely. Thanks!

**Pull request merged**